### PR TITLE
expression: function SUBSTR/LEFT/RIGHT should manipulate code points instead of bytes(#4019)

### PR DESCRIPTION
**Pull request by zz-jason**
fix https://github.com/pingcap/tidb/issues/4019

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @shenli PTAL

File: expression/builtin_string.go [hunk 1]
```diff
@@ -350,13 +350,15 @@ func (b *builtinLeftSig) evalString(row []types.Datum) (d string, isNull bool, e
 	if isNull || err != nil {
 		return d, isNull, errors.Trace(err)
 	}
+	runes := []rune(d)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: length := len([]rune(d))

File: expression/builtin_string.go [hunk 2]
```diff
@@ -380,25 +382,24 @@ type builtinRightSig struct {
...
 	}
-	right, isNull, err = b.args[1].EvalInt(row, sc)
+	right, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
 		return d, isNull, errors.Trace(err)
 	}
 	r := int(right)
-	length := len(d)
+	runes := []rune(d)
+	numRunes := len(runes)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 3]
```diff
@@ -781,30 +782,26 @@ type builtinSubstring2ArgsSig struct {
...
 		return d, isNull, errors.Trace(err)
 	}
-	pos, isNull, err = b.args[1].EvalInt(row, sc)
+	pos, isNull, err := b.args[1].EvalInt(row, sc)
 	if isNull || err != nil {
 		return d, isNull, errors.Trace(err)
 	}
-	strLen := int64(len(str))
+	runes := []rune(str)
+	numRunes := int64(len(runes))
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

File: expression/builtin_string.go [hunk 4]
```diff
@@ -814,40 +811,36 @@ type builtinSubstring3ArgsSig struct {
...
 	if isNull || err != nil {
 		return d, isNull, errors.Trace(err)
 	}
-	length, isNull, err = b.args[2].EvalInt(row, sc)
+	length, isNull, err := b.args[2].EvalInt(row, sc)
 	if isNull || err != nil {
 		return d, isNull, errors.Trace(err)
 	}
-	strLen := int64(len(str))
+	runes := []rune(str)
```
Comment (XuHuaiyu) [inline expression/builtin_string.go]: ditto

**Comment by XuHuaiyu** [comment]
LGTM

**Pull request merged**